### PR TITLE
Enable new compiler diagnostics for invalid array accesses

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

**Review by yuwata — APPROVED** [review]

File: src/fundamental/macro-fundamental.h [hunk 1]
```diff
@@ -376,3 +376,15 @@ static inline size_t ALIGN_TO(size_t l, size_t ali) {
...
+
+/* Declare a flexible array usable in a union.
+ * This is essentially a work-around for a pointless constraint in C99
+ * and might go away in some future version of the standard.
+ *
+ * See https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=3080ea5553cc909b000d1f1d964a9041962f2c5b
+ */
+#define DECLARE_FLEX_ARRAY(type, name)                 \
+        struct {                                       \
+                struct {} __empty__ ## name;           \
```
Comment (poettering) [inline src/fundamental/macro-fundamental.h]: maybe use the existing `dummy_t` for this?

File: src/libsystemd/sd-journal/journal-def.h [hunk 2]
```diff
@@ -93,22 +93,28 @@ struct FieldObject FieldObject__contents;
...
+#define EntryObject__contents {                        \
+        ObjectHeader object;                           \
+        le64_t seqnum;                                 \
+        le64_t realtime;                               \
+        le64_t monotonic;                              \
+        sd_id128_t boot_id;                            \
+        le64_t xor_hash;                               \
+        union {                                        \
+                struct {                               \
+                        struct {} __empty__regular;    \
```
Comment (poettering) [inline src/libsystemd/sd-journal/journal-def.h]: maybe use dummy_t?

File: src/libsystemd/sd-journal/journal-def.h [hunk 3]
```diff
@@ -93,22 +93,28 @@ struct FieldObject FieldObject__contents;
...
+                        struct {                       \
+                                le64_t object_offset;  \
+                                le64_t hash;           \
+                        } regular[];                   \
+                };                                     \
+                struct {                               \
+                        struct {} __empty_compact;     \
+                        struct {                       \
+                                le32_t object_offset;  \
+                        }  compact[];                  \
```
Comment (poettering) [inline src/libsystemd/sd-journal/journal-def.h]: spurious whitespace?

File: src/libsystemd/sd-journal/journal-def.h [hunk 4]
```diff
@@ -93,22 +93,28 @@ struct FieldObject FieldObject__contents;
...
+        union {                                        \
+                struct {                               \
+                        struct {} __empty__regular;    \
+                        struct {                       \
+                                le64_t object_offset;  \
+                                le64_t hash;           \
+                        } regular[];                   \
+                };                                     \
+                struct {                               \
+                        struct {} __empty_compact;     \
```
Comment (poettering) [inline src/libsystemd/sd-journal/journal-def.h]: dummy_t

File: src/fundamental/macro-fundamental.h [hunk 5]
```diff
@@ -376,3 +376,15 @@ static inline size_t ALIGN_TO(size_t l, size_t ali) {
         (v) = UPDATE_FLAG(v, flag, b)
 #define FLAGS_SET(v, flags) \
         ((~(v) & (flags)) == 0)
```
Comment (poettering) [inline src/fundamental/macro-fundamental.h:378]: why would the macro only work for basic types? i don't get this? i mean, the macro would just expand to the right stuff either way? can you explain?

Comment (poettering) [reply src/fundamental/macro-fundamental.h:378]: and i guess structs are actually fine, but anonymous ones are not?

Comment (keszybz) [reply src/fundamental/macro-fundamental.h:378]: Yeah, anonymous structure is a problem. We *could* make it non-anonymous, but I think we don't want to do that. I'll adjust the commit message.

**Comment by keszybz** [comment]
Updated.

In the previous version, I missed an spot in `shared/json.h`.  It is adjusted now too.

**Pull request merged**